internal/coreinternal/golden/sort_attributes.go

```diff
@@ -0,0 +1,96 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package golden // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/golden"
+
+import (
+	"fmt"
+	"sort"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+// sorts all Resource Metrics attributes and Datapoint Slice metric attributes
+func sortMetrics(ms pmetric.Metrics) {
+	rms := ms.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		sortAttributeMap(rms.At(i).Resource().Attributes())
+		ilms := rms.At(i).ScopeMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			sortAttributeMap(ilms.At(j).Scope().Attributes())
+			metricsList := ilms.At(j).Metrics()
+			for k := 0; k < metricsList.Len(); k++ {
+				metric := metricsList.At(k)
+				switch metricsList.At(k).Type() {
+				case pmetric.MetricTypeGauge, pmetric.MetricTypeSum:
+					ds := getDataPointSlice(metric)
```

Comment:
I think the function `getDataPointSlice` may be not necessary? You can just change this part into following
```go
switch metricsList.At(k).Type() {
	case pmetric.MetricTypeGauge:
		ds = metric.Gauge().DataPoints()
	case pmetric.MetricTypeSum:
		ds = metric.Sum().DataPoints()
        case pmetric.MetricTypeHistogram:
                 ds = metric.Histogram().DataPoints()
        case pmetric.MetricTypeExponentialHistogram:
                 ds = metric.ExponentialHistogram().DataPoints()
        case pmetric.MetricTypeSummary:
                 ds = metric.Summary().DataPoints()
}
for l := 0; l < ds.Len(); l++ {
	sortAttributeMap(ds.At(l).Attributes())
}
```

please correct me if I misunderstand this

Reply:
:+1 Good catch.

Reply:
I agree with removing the function, but i believe i still need the for-loops with calls to sortAttributeMap within their individual case statements due to scoping of the variable ds.

Reply:
It's true that most of these cases return different types from `DataPoints` - `NumberDataPointSlice`, `HistogramDataPointSlice`, `SummaryDataPointSlice`, so I think the for loop in each case is fine.